---
title: "Project 2D || TNFα (Tumor Necrosis Factor - alpha) protein"
author: "Balang Qasab"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
    highlight: tango
---

## Introduction

### Scientific Question:

Can a model organism for melanoma drug treatment be identified via high sequence similarity to TNFα protein in humans?

### Hypothesis:

If an organism's TNFα protein sequence is <70% similarity to Human TNFα protein sequence, especially near the ligand binding site, then that organism can be potentially used a model organism for drug treatment testing.

### Background:

TNF (Tumor necrosis factor) is also known as TNF-α (tumor necrosis factor alpha). It is a protein consisting of about 157 amino acids and weighs 17kDa. It also acts as an homotimer in a solution which is largely generated by T lymphocytes, activated macrophages and NK (natural killer) cells. [1]

TNF-α is both a cytokine and an adipokine. It is utilized by immune system for cell signaling for example if an infection is detected by macrophages, they naturally release TNF to alert other cells in immune system and create an inflammatory response. This was the cytokine role of TNF-α and as a role of adipokine, it promotes insulin resistance.

As mentioned above that TNF-α is natural product of activated macrophages and monocytes it also exhibits pleiotropic effects on the both normal and malignant cells. [3] This means that the effect of TNF - alpha associated inflammation may actually be cancer promoting. Tumor Necrosis Factor-α inhibitors (TNFαIs) are monoclonal antibodies commonly used in the treatment of many dermatologic, rheumatologic, and gastrointestinal inflammatory disorders. 

Melanoma is an aggressive type of skin cancer which develops in melanin producing cells called melanocytes. Melanin is the pigment that provided color to the skin. The cause of melanoma is still not very clear but UV radiation exposure from sunglight is found to increase the risk of developing melanoma. Limiting exposure to same can reduce the risk. 
It is treatable cancer if early detection is done. [4]

In a 2013 study [5], the researchers suggested association between TNFαIs and melanoma after a RADAR project (Research on Adverse Drug events And Reports). This study suggested use of TNFαIs might be causing malignant melanoma. On the contrast this and other similar research, a study published in 2015 author demonstrated that chronic treatment with TNFαIs does not increase the risk of developing melanoma. [6] 

Despite the bidirectional relation of TNF-aplha inhibitors and melanoma. TNF-alpha is a commonly studied protein for cancer treatment. [7]. In this bioinformatics analysis I try to identify a model organism for melanoma drug treatment target TNF-alpha protein. 

1. F. Atzeni, P. Sarzi-Puttini, in Brenner's Encyclopedia of Genetics (Second Edition), 2013
2. Sethi JK, Hotamisligil GS (2021). "Metabolic Messengers: tumour necrosis factor". Nature Metabolism. 3 (10): 1302–1312. 
3. Meyler's Side Effects of Drugs (Sixteenth Edition), 2016
4. https://www.mayoclinic.org/diseases-conditions/melanoma/symptoms-causes/syc-20374884
5. Nardone, B., Hammel, J., Raisch, D., Weaver, L., Schneider, D. and West, D. (2014), Melanoma associated with tumour necrosis factor-α inhibitors: a Research on Adverse Drug events And Reports (RADAR) project. Br J Dermatol, 170: 1170-1172. https://doi.org/10.1111/bjd.12779
6. Donia, M., Kjeldsen, J. W., & Svane, I. M. (2015). The controversial role of TNF in melanoma. Oncoimmunology, 5(4), e1107699. https://doi.org/10.1080/2162402X.2015.1107699
7. Van Horssen, R., Ten Hagen, T. L., & Eggermont, A. M. (2006). TNF-alpha in cancer treatment: molecular insights, antitumor effects, and clinical utility. The oncologist, 11(4), 397–408. https://doi.org/10.1634/theoncologist.11-4-397


### Project Description

In this project, I performed pairwise sequence alignment and multiple sequence alignment to identify similarities in the 10 selected sequences of TNF-alpha protein downloaded from Uniprot. The results were visualized using MSA output in the notebook, heatmap of pairwise alignment score and sequence logo visualization for MSA. 

Further I tried to identify residues in binding sites of TNF-alpha protein. The protein was downloaded from rcsb.org and PDB ID is 5VH4. I used bio3d package to find residue number in the binding site of protein and displayed them in my notebook. Further I also used NGLViewer to visualize the protein structure in the notebook.  

## Package Descriptions

```{r loading pacages, message=FALSE, warning=FALSE}
# Install CRAN packages using install.packages() and Bioconductor package using BiocManager::install()

# To read the FASTA sequence data and perform pairwise alignment
library(Biostrings) 

# To perform multiple sequence alignment 
library(msa) 

# To identify residues in protein binding site
library(bio3d)
library(gplots)

# To convert tex format to pdf after MSA results
library(tools)

# To visualize protein structure
library(NGLVieweR)

```

## Performing Bioinformatics Analyses

### Reading the FASTA file and defining GLOBAL variable and functions

```{r}
# TNFA_uniprot is an AAStringSet variable or an S4 variable that is a global variable
TNFA_uniprot <- readAAStringSet("TNFA_aa_seq.fasta") # Read FASTA (or FASTQ) files in an XStringSet object:

# to check datatype
typeof(TNFA_uniprot)

# to check length of s4 object that will tell how many FASTA sequences it contains.
length(TNFA_uniprot)

names(TNFA_uniprot) # Function to get the names of an object.

# Defining a function used later for pairwise alignment
matrix4heatmap <- function(S4sequences) {

  # Store length/no. of sequences in x
  x <- length(S4sequences)
  # Create a grid of possible combinations
  seq_index = expand.grid(1:x, 1:x)


  # Use apply function to iterate pairwise alignment score calculation on possible combinations of sequences
  forheatmap  <- apply(seq_index,1, function(x){
    
    seq1 <- S4sequences[x[1]]
    seq2 <- S4sequences[x[2]]
    # Turn the seq1 variable into a character string
    
    dsl = AAStringSetList(seq1)
    as.character(unlist(dsl))
    cl = as(dsl, "CharacterList")
    as.list(cl)
    
    # Turn the seq2 variable into a character string
    dsl2 = AAStringSetList(seq2)
    as.character(unlist(dsl2))
    cl2 = as(dsl2, "CharacterList")
    as.list(cl2)
    
    # Store the score in variable s
    s <- score(pairwiseAlignment(pattern = c(cl2), subject = cl))
    
    # return s
    return(s)
  })
  
  forheatmap <- matrix(forheatmap, nrow=x, ncol=x)

  # Return the generated score matrix 
  return(forheatmap)
}



```


### Perform Pairwise Sequence Alignment and generate a heatmap

```{r}
# pairwise_score_mat stores the scoring matrix from matrix4heatmap function. 
pairwise_score_mat <- matrix4heatmap(TNFA_uniprot)

# Assign row and column names to matrix to display in heatmap  
colnames(pairwise_score_mat) <- substr(names(TNFA_uniprot), 11,20) # Use a substr to show only specific string length
rownames(pairwise_score_mat) <- substr(names(TNFA_uniprot), 11,20)

# Generate and display heatmap with custom margins and tracecol as NA. 
heatmap.2(pairwise_score_mat,margin=c(10, 10), tracecol=NA)

```

### Identification of residue in binding sites using bio3d

```{r message=FALSE, warning=FALSE}

# 5VH4 : Crystal structure of Fab fragment of anti-TNFa antibody infliximab in an I-centered orthorhombic crystal form

# Read in the pdb
pdb_5VH4 <- read.pdb('5VH4') # read.pdb(): Read a Protein Data Bank (PDB) coordinate file.

## Automatically identify what is the protein and what is the ligand
bs <- binding.site(pdb_5VH4) # binding.site() : Determines the interacting residues between two PDB entities.

# Print the residues & save the resnumbers to a vector for later use
bs_resnumbers <- bs$resno

bs$resnames


```

## Visualization and exploration of protein structure

```{r}
# Use NGLViewer to fetch and display protein structure.
# Custom visual parameters are applied to make visualization better.

NGLVieweR("5VH4") %>%
  stageParameters(backgroundColor = "black") %>%
  setQuality("high") %>%
  addRepresentation("cartoon", param = list(name = "cartoon",
                                            colorScheme = "residueindex",
                                            showBackground = TRUE,
                                            labelType = "res",
                                            backgroundColor = "white",
                                            backgroundOpacity = 0.8))

```


### Perform Multiple Sequence Alignment

```{r message=FALSE, warning=FALSE}

TNFA_MSA <- msa(TNFA_uniprot) # Multiple Sequence Alignment

print(TNFA_MSA, show="complete")

```


### Convert results of MSA to Latex and PDF output for visualization

```{r fig.height=9, fig.width=12, message=FALSE, warning=FALSE}

# Code check below for what the first value is in the bs_resnumbers vector and how long that vector is.
bs_resnumbers[1]

length(bs_resnumbers)

# msaPrettyPrint is a function in the msa package. First you pass it the multiple sequence alignment output from
# above, TNFA_MSA Then you specify the output type, and tex stands for latex. There are many other types you 
# could do but this is the one we are working with here so that we can create a PDF ultimately.
# Y specifies the residue sequence range to look at with the image. 
## create PDF file according to some custom settings

tex_file <- paste(getwd(),"/msa_output.tex",sep="")

msaPrettyPrint(TNFA_MSA,  file = tex_file, output="tex", y=c((bs_resnumbers[1]+35), (bs_resnumbers[length(bs_resnumbers)]+35)), showNames="left", showLogo="none",
logoColors="rasmol", shadingMode="functional", shadingModeArg="structure",
verbose = TRUE, showLegend=FALSE, askForOverwrite = FALSE)

# texi2pdf(): Run latex/pdflatex, makeindex and bibtex until all cross-references are resolved to create a dvi or a PDF file.
texi2pdf(tex_file, clean=TRUE) 


```

### Display MSA results from PDF output.

![MSA results](./msa_output.pdf){width=100%}



## Analysis of Results


* From the pairwise alignment results we can observe a high alignment score of human TNF-alpha protein with that of DOG, HORSE and PIG. The highest alignment score of of sequence with that of HUMAN is TNFA_CANLF Tumor necrosis factor OS=Canis lupus familiaris with UniProt ID: P51742. 

* From visualization of the protein with PDB: 5VH4 and name: ID Crystal structure of Fab fragment of anti-TNFa antibody infliximab in an I-centered orthorhombic crystal form; we can observe it has 2 amino chains.

* The results of MSA output showed TNF-alpha protein sequence of SHEEP, CAPHI and, BOVIN have highest alignment score.The sequence HUMAN is present in the 6th position with 5th as CANLF and 7th as HORSE. This was also confirmed during pairwise sequence alignment the CANLF and HORSE had highest pairwise alignment score with HUMAN sequence.

The final results of analysis indicated CANLF (Dog) and HORSE (Horse) to be potentially used as model organism for melanoma drug treatment (or other cancer as well) with target of TNF-alpha protein.